### PR TITLE
[SPARK-10082] [MLlib] Validate i, j in apply DenseMatrices and SparseMatrices

### PR DESCRIPTION
**Pull request by MechCoder**
Given row_ind should be less than the number of rows
Given col_ind should be less than the number of cols.

The current code in master gives unpredictable behavior for such cases.

### CODE REVIEW

**Comment by MechCoder** [comment]
10k JIRA's wohoo :D

**Comment by feynmanliang** [comment]
:+1: at first I was concerned that this might be unnecessary since `Array` checks for bounds, but this PR fixes a bug which occurs when a user provides an OOB row index and `values` causes overflow to the next column (assuming row-major order)

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 1]
```diff
@@ -281,7 +281,11 @@ class DenseMatrix(
   /**
    * @since 1.3.0
    */
-  override def apply(i: Int, j: Int): Double = values(index(i, j))
+  override def apply(i: Int, j: Int): Double = {
+    require(i < numRows && i >=0, s"Expected i to be non-negative and < $numRows, got $i")
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: nit: "Expected 0 <= i < $numRows bot got i=$i", same for others

Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Actually, perhaps we should move this check into `index` so other methods (e.g. `update`) will also get checked

File: mllib/src/test/scala/org/apache/spark/mllib/linalg/MatricesSuite.scala [hunk 2]
```diff
@@ -74,6 +74,17 @@ class MatricesSuite extends SparkFunSuite {
     }
   }
 
+  test("apply in matrices incorrect input") {
```
Comment (feynmanliang) [inline mllib/src/test/scala/org/apache/spark/mllib/linalg/MatricesSuite.scala]: After moving check to `index`, check `index` here instead of `apply`

**Comment by MechCoder** [comment]
Actually this check is not necessary for SparseMatrices. But worst case it saves the effort of doing a binary search (and it should not return 0.0 if the row and col is not present)

I have addressed your comments/

**Comment by MechCoder** [comment]
And I discovered the OOB bug by accident :)

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 3]
```diff
@@ -281,9 +281,13 @@ class DenseMatrix(
   /**
    * @since 1.3.0
    */
-  override def apply(i: Int, j: Int): Double = values(index(i, j))
+  override def apply(i: Int, j: Int): Double = {
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: nit: this is a noop change, can we just keep the previous code?

**Comment by feynmanliang** [comment]
minor suggestion but LGTM

CC @mengxr @jkbradley for final pass

**Comment by mengxr** [comment]
I think this requires some micro-benchmark. I want to see the overhead of additional two checks. We can also test a single `require` statement that contains both checks.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 4]
```diff
@@ -284,6 +284,8 @@ class DenseMatrix(
   override def apply(i: Int, j: Int): Double = values(index(i, j))
 
   private[mllib] def index(i: Int, j: Int): Int = {
+    require(i < numRows && i >=0, s"Expected 0 <= i < $numRows, got $i")
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Nit: space after `>=`, and yes I think these can be combined into one check each. If there's an error I think it's helpful to log both i and j anyway.

**Comment by mengxr** [comment]
Merged into master. I will send a follow-up PR to address minor issues.

**Comment by MechCoder** [comment]
thanks !

**Pull request closed without merging**